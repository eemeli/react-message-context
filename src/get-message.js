function getIn(messages, path) {
  if (!messages) return undefined
  let msg = messages
  for (let i = 0; i < path.length; ++i) {
    msg = msg[path[i]]
    if (msg === undefined) return undefined
  }
  return msg
}

export function getPath(id, pathSep) {
  if (!id) return []
  if (Array.isArray(id)) return id
  return pathSep ? id.split(pathSep) : [id]
}

/**
 * @private
 * @param {Context} context
 * @param {string|string[]} id
 * @param {string|string[]} [locale]
 * @returns {any}
 */
export default function getMessage({ locales, messages, pathSep }, id, locale) {
  if (locale != null) locales = Array.isArray(locale) ? locale : [locale]
  const path = getPath(id, pathSep)
  for (let i = 0; i < locales.length; ++i) {
    const lc = locales[i]
    const msg = getIn(messages[lc], path)
    if (msg !== undefined) return msg
  }
  return undefined
}
