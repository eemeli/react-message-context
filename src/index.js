export { default as Message } from './message'
export { default as MessageProvider } from './message-provider'
export { default as useLocales } from './use-locales'
export { default as useMessageGetter } from './use-message-getter'
export { default as withMessages } from './with-messages'
