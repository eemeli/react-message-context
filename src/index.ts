export { getMessage, getMessageGetter } from './get-message'
export { Message, MessageProps } from './message'
export { MessageContext } from './message-context'
export { MessageError } from './message-error'
export { MessageProvider, MessageProviderProps } from './message-provider'
export { Id, MessageObject, MessageValue } from './types'
export { useLocales } from './use-locales'
export { useMessage } from './use-message'
export { useMessageGetter } from './use-message-getter'
