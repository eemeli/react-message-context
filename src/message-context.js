import React from 'react'

const defaultValue = {
  locales: [],
  messages: {},
  pathSep: undefined
}
const context = React.createContext(defaultValue)
const { Consumer, Provider } = context

export { context as default, Consumer, Provider, defaultValue }
