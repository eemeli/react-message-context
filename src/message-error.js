export const errorMessages = {
  EBADMSG: 'Message with unexpected object value',
  ENOMSG: 'Message not found'
}

export class MessageError extends Error {
  constructor(path, code, asId) {
    super(`${errorMessages[code]}: ${asId(path)}`)
    this.code = code
    this.path = path
  }
}

