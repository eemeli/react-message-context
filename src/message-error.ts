export const errorMessages = {
  EBADMSG: 'Message with unexpected object value',
  ENOMSG: 'Message not found'
}

export type ErrorCode = keyof typeof errorMessages

export class MessageError extends Error {
  code: ErrorCode
  path: string[]

  constructor(
    path: string[],
    code: ErrorCode,
    asId: (path: string[]) => string
  ) {
    super(`${errorMessages[code]}: ${asId(path)}`)
    this.code = code
    this.path = path
  }
}
