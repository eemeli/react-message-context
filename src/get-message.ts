import { MessageContext } from './message-context'
import { Id, MessageObject, MessageValue } from './types'

function getIn(messages: MessageValue | MessageObject, path: string[]) {
  if (messages) {
    for (let i = 0; i < path.length; ++i) {
      if (typeof messages !== 'object') return undefined
      messages = messages[path[i]]
      if (messages === undefined) return undefined
    }
  }
  return messages
}

export function getPath(id?: Id, pathSep?: string | null) {
  if (!id) return []
  if (Array.isArray(id)) return id
  return pathSep ? id.split(pathSep) : [id]
}

/**
 * Given a `MessageContext` instance, fetches an entry from the messages object of the current or given locale.
 * The returned value will be `undefined` if not found, or otherwise exactly as set in the `MessageProvider` props.
 *
 * @public
 * @param context - The `MessageContext` instance
 * @param id - The key or key path of the message or message object.
 *   If empty or `[]`, matches the root of the messages object
 * @param locale - If set, overrides the current locale precedence as set by parent MessageProviders.
 */
export function getMessage(
  { locales, messages, onError, pathSep }: MessageContext,
  id?: Id,
  locale?: Id
) {
  if (locale != null) locales = Array.isArray(locale) ? locale : [locale]
  const path = getPath(id, pathSep)
  for (let i = 0; i < locales.length; ++i) {
    const lc = locales[i]
    const msg = getIn(messages[lc], path)
    if (msg !== undefined) return msg
  }
  return onError ? onError(path, 'ENOMSG') : undefined
}

/**
 * @param id - Message identifier; extends the path set by `rootId`
 * @param params - Parameters for a function message
 */
export interface MessageGetterOptions {
  baseParams?: any
  locale?: Id
}

/**
 * Given a `MessageContext` instance, returns a message getter function, which may have a preset root id path, locale, and/or base parameters for message functions.
 *
 * The returned function takes two parameters `(msgId, msgParams)`, which will extend any values set by the hook's arguments.
 *
 * @public
 * @param context - The `MessageContext` instance
 * @param rootId - The key or key path of the message or message object.
 *   If empty or `[]`, matches the root of the messages object
 * @param options - If `baseParams` is set, message function parameters will be assumed to always be an object, with these values initially set.
 *   `locale` overrides the current locale precedence as set by parent MessageProviders.
 */
export function getMessageGetter(
  context: MessageContext,
  rootId?: Id,
  { baseParams, locale }: MessageGetterOptions = {}
) {
  const { pathSep } = context
  const pathPrefix = getPath(rootId, pathSep)
  return function message(id: Id, params: any) {
    const path = pathPrefix.concat(getPath(id, pathSep))
    const msg = getMessage(context, path, locale)
    if (typeof msg !== 'function') return msg
    const msgParams = baseParams
      ? Object.assign({}, baseParams, params)
      : params
    return msg(msgParams)
  }
}
