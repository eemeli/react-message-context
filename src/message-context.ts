import { Context /* used in d.ts */, createContext } from 'react'
import { ErrorCode } from './message-error'

/** @internal */
export type MessageValue = string | number | boolean | ((props: any) => any)

/** @internal */
export interface MessageObject {
  [key: string]: MessageValue | MessageObject
}

/** @public */
export interface MessageContext {
  locales: string[]
  merge: (target: MessageObject, ...sources: MessageObject[]) => MessageObject
  messages: MessageObject

  /** Always defined in MessageProvider children */
  onError?: (path: string[], code: ErrorCode) => any
  pathSep: string | null
}

export const defaultValue: MessageContext = {
  locales: [],
  merge: Object.assign,
  messages: {},
  pathSep: '.'
}

/**
 * The context object used internally by the library.
 * Probably only useful with `Class.contextType` or for building custom hooks.
 *
 * @public
 *
 * @example
 * ```js
 * import React, { Component } from 'react'
 * import {
 *   getMessage,
 *   getMessageGetter,
 *   MessageContext,
 *   MessageProvider
 * } from 'react-message-context'
 *
 * const messages = {
 *   example: { key: 'Your message here' },
 *   other: { key: 'Another message' }
 * }
 *
 * class Example extends Component {
 *   render() {
 *     const message = getMessage(this.context, 'example.key')
 *     const otherMsg = getMessageGetter(this.context, 'other')
 *     return (
 *       <span>
 *         {message} | {otherMsg('key')}
 *       </span>
 *     ) // 'Your message here | Another message'
 *   }
 * }
 * Example.contextType = MessageContext
 *
 * export const App = () => (
 *   <MessageProvider messages={messages}>
 *     <Example />
 *   </MessageProvider>
 * )
 * ```
 */
export const MessageContext = createContext(defaultValue)
