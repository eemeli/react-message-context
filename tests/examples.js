import { fromJS } from 'immutable'
import React, { Component } from 'react'
import renderer from 'react-test-renderer'

import { Message, MessageProvider, withMessages } from '../src/index'

// actually precompiled with messageformat-cli
import en from './fixtures/messages_en'
import fi from './fixtures/messages_fi'

test('Example 1', () => {
  const messages = {
    message: 'Your message',
    answers: {
      sixByNine: ({ base }) => (6 * 9).toString(base),
      universe: 42
    }
  }

  const Equality = ({ messages }) => {
    const foo = messages('sixByNine', { base: 13 })
    const bar = messages('universe')
    return `${foo} and ${bar} are equal`
  }
  const WrappedEquality = withMessages('answers')(Equality)

  const App = () => (
    <ul>
      <li>
        <Message id="message" /> is important
      </li>
      <li>
        The answer is <Message id="answers.sixByNine" base={13} />
      </li>
      <li>
        <WrappedEquality />
      </li>
    </ul>
  )

  const component = renderer.create(
    <MessageProvider messages={messages}>
      <App />
    </MessageProvider>
  )
  expect(component.toJSON()).toMatchSnapshot()
})

describe('Example 2', () => {
  const Errors = () => (
    <ul>
      <li>
        <Message id="errors.wrong_length" length={42} />
      </li>
      <li>
        <Message id="errors.equal_to" count={13} />
      </li>
    </ul>
  )

  test('With plain object', () => {
    const component = renderer.create(
      <MessageProvider locale="en" messages={en}>
        <MessageProvider locale="fi" messages={fi}>
          <Errors />
        </MessageProvider>
      </MessageProvider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('With immutable Map', () => {
    const component = renderer.create(
      <MessageProvider locale="en" messages={fromJS(en)}>
        <MessageProvider locale="fi" messages={fromJS(fi)}>
          <Errors />
        </MessageProvider>
      </MessageProvider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('Example 3', () => {
  test('With plain object', () => {
    const component = renderer.create(
      <MessageProvider locale="en" messages={{ foo: 'FOO', qux: 'QUX' }}>
        <MessageProvider locale="fi" messages={{ foo: 'FÖÖ', bar: 'BÄR' }}>
          <Message id="foo" />
          <Message id="foo" locale="en" />
          <Message id="bar" />
          <Message id="bar" locale="en" />
          <Message id="qux" />
        </MessageProvider>
      </MessageProvider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('With immutable Map', () => {
    const en = fromJS({ foo: 'FOO', qux: 'QUX' })
    const fi = fromJS({ foo: 'FÖÖ', bar: 'BÄR' })
    const component = renderer.create(
      <MessageProvider locale="en" messages={en}>
        <MessageProvider locale="fi" messages={fi}>
          <Message id="foo" />
          <Message id="foo" locale="en" />
          <Message id="bar" />
          <Message id="bar" locale="en" />
          <Message id="qux" />
        </MessageProvider>
      </MessageProvider>
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})
