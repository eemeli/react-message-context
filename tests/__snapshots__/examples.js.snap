// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example 1 1`] = `
<ul>
  <li>
    Your message
     is important
  </li>
  <li>
    The answer is 
    42
  </li>
  <li>
    42 and 42 are equal
  </li>
</ul>
`;

exports[`Example 2 With immutable Map 1`] = `
<ul>
  <li>
    Viestisi on väärän pituinen (pitäisi olla 42 merkkiä)
  </li>
  <li>
    The value must be equal to 13
  </li>
</ul>
`;

exports[`Example 2 With plain object 1`] = `
<ul>
  <li>
    Viestisi on väärän pituinen (pitäisi olla 42 merkkiä)
  </li>
  <li>
    The value must be equal to 13
  </li>
</ul>
`;

exports[`Example 3 With immutable Map 1`] = `
Array [
  "FÖÖ",
  "FOO",
  "BÄR",
  "bar",
  "QUX",
]
`;

exports[`Example 3 With plain object 1`] = `
Array [
  "FÖÖ",
  "FOO",
  "BÄR",
  "bar",
  "QUX",
]
`;
